Assistance with exception specification errors

Hi, I'm looking into the jemalloc package in Nixpkgs failing to build for me on macOS.

For reference, [the package expression](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/pkgs/development/libraries/jemalloc/common.nix).

Here's the build errors I get:

```
In file included from src/jemalloc_cpp.cpp:9:            
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:         
include/jemalloc/internal/../jemalloc.h:226:28: error: exception specification in declaration does not match previous declaration
    void JEMALLOC_NOTHROW       *je_malloc(size_t size)
                                 ^                                              
include/jemalloc/internal/../jemalloc.h:69:21: note: expanded from macro 'je_malloc' 
#  define je_malloc malloc            
                    ^           
/nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/stdlib.h:152:7: note: previous declaration is here
void    *malloc(size_t __size) __result_use_check;
         ^                                                                      
In file included from src/jemalloc_cpp.cpp:9:            
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:         
include/jemalloc/internal/../jemalloc.h:229:28: error: exception specification in declaration does not match previous declaration
    void JEMALLOC_NOTHROW       *je_calloc(size_t num, size_t size)
                                 ^                                              
include/jemalloc/internal/../jemalloc.h:63:21: note: expanded from macro 'je_calloc'
#  define je_calloc calloc            
                    ^           
/nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/stdlib.h:140:7: note: previous declaration is here
 void    *calloc(size_t __count, size_t __size) __result_use_check;
          ^                                                                      
In file included from src/jemalloc_cpp.cpp:9:    
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:         
include/jemalloc/internal/../jemalloc.h:231:38: error: exception specification in declaration does not match previous declaration
JEMALLOC_EXPORT int JEMALLOC_NOTHROW    je_posix_memalign(void **memptr,
                                        ^                                       
include/jemalloc/internal/../jemalloc.h:77:29: note: expanded from macro 'je_posix_memalign'
#  define je_posix_memalign posix_memalign
                            ^   
/nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/stdlib.h:156:7: note: previous declaration is here
int      posix_memalign(void **__memptr, size_t __alignment, size_t __size) __OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_3_0);                          
         ^                                                                      
In file included from src/jemalloc_cpp.cpp:9:                                   
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:
include/jemalloc/internal/../jemalloc.h:238:28: error: exception specification in declaration does not match previous declaration
    void JEMALLOC_NOTHROW       *je_realloc(void *ptr, size_t size)
                                 ^                                              
include/jemalloc/internal/../jemalloc.h:79:22: note: expanded from macro 'je_realloc'                                   
#  define je_realloc realloc
                     ^
/nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/stdlib.h:160:7: note: previous declaration is here              
void    *realloc(void *__ptr, size_t __size) __result_use_check;
         ^
 In file included from src/jemalloc_cpp.cpp:9:                                   
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:
include/jemalloc/internal/../jemalloc.h:240:39: error: exception specification in declaration does not match previous declaration
JEMALLOC_EXPORT void JEMALLOC_NOTHROW   je_free(void *ptr)
                                        ^
include/jemalloc/internal/../jemalloc.h:65:19: note: expanded from macro 'je_free'
#  define je_free free
                  ^
 /nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/stdlib.h:143:7: note: previous declaration is here
void     free(void *);
         ^
In file included from src/jemalloc_cpp.cpp:9:
In file included from include/jemalloc/internal/jemalloc_preamble.h:21:
include/jemalloc/internal/../jemalloc.h:279:28: error: exception specification in declaration does not match previous declaration
    void JEMALLOC_NOTHROW       *je_valloc(size_t size) JEMALLOC_CXX_THROW
                                 ^
include/jemalloc/internal/../jemalloc.h:83:21: note: expanded from macro 'je_valloc'
#  define je_valloc valloc
                    ^
/nix/store/qqkps7h1yg2fy2bf482hbzcy5ava3pys-Libsystem-1238.60.2/include/unistd.h:726:7: note: previous declaration is here
void    *valloc(size_t);                         
         ^
```

I suspect the problem may be the somewhat older version of Libsystem in Nixpkgs. I'm working on updating this but it's a slow process. Does this sound like a likely problem and if so, is there a simple patch we could apply until we get Libsystem updated, could we simply drop all the `JEMALLOC_NOTHROW` for example?

--- a/Makefile.in
+++ b/Makefile.in
@@ -458,6 +458,8 @@ $(TESTS_OBJS): $(objroot)test/%.$(O): $(srcroot)test/%.c
 $(TESTS_CPP_OBJS): $(objroot)test/%.$(O): $(srcroot)test/%.cpp
 $(TESTS_OBJS): CPPFLAGS += -I$(srcroot)test/include -I$(objroot)test/include
 $(TESTS_CPP_OBJS): CPPFLAGS += -I$(srcroot)test/include -I$(objroot)test/include
+$(TESTS_OBJS): CFLAGS += -fno-builtin
+$(TESTS_CPP_OBJS): CPPFLAGS += -fno-builtin
 ifneq ($(IMPORTLIB),$(SO))
 $(CPP_OBJS) $(C_SYM_OBJS) $(C_OBJS) $(C_JET_SYM_OBJS) $(C_JET_OBJS): CPPFLAGS += -DDLLEXPORT
 endif

